File: src/results/atom/legend/Legend.java
package org.treez.results.atom.legend;

import java.util.List;
import java.util.Objects;

import org.apache.log4j.Logger;
import org.eclipse.swt.graphics.Image;
import org.treez.core.adaptable.FocusChangingRefreshable;
import org.treez.core.atom.graphics.GraphicsPropertiesPageFactory;
import org.treez.core.treeview.treeView;
import org.treez.javafxd3.d3.D3;
import org.treez.javafxd3.d3.core.Selection;
import org.treez.results.Activator;
import org.treez.results.atom.graphicsPage.Background;
import org.treez.results.atom.graphicsPage.Border;
import org.treez.results.atom.graphicsPage.GraphicsPropertiesPage;

@SuppressWarnings("checkstyle:visibilitymodifier")
public class Legend extends GraphicsPropertiesPage {

	@SuppressWarnings("unused")
	private static final Logger LOG = Logger.getLogger(Legend.class);

	//#region ATTRIBUTES

	public Main main;

	public Text text;

	public Background background;

	public Border border;

	private Selection legendGroupSelection;

	private Selection rectSelection;

	//#end region

	//#region CONSTRUCTORS

	public Legend(String name) {
		super(name);
	}

	//#end region

	//#region METHODS

	@Override
	protected void createPropertyPageFactories() {

		main = new Main();
		propertyPageFactories.add(main);

		text = new Text();
		propertyPageFactories.add(text);

		background = new Background();
		propertyPageFactories.add(background);

		border = new Border();
		propertyPageFactories.add(border);

	}

	@Override
	public Image provideImage() {
		return Activator.getImage("legend.png");
	}

	@Override
	protected List<Object> extendContextMenuActions(List<Object> actions, treeView treeViewer) {
		// no actions available right now
		return actions;
	}

	@Override
	public Selection plotWithD3(
			D3 d3,
			Selection graphSelection,
			Selection graphRectSelection,
			FocusChangingRefreshable refreshable) {
		Objects.requireNonNull(d3);
		this.treeView = refreshable;

		graphSelection //
				.select("#" + name) //
				.remove();

		legendGroupSelection = graphSelection //
				.append("g") //
				.attr("class", "legend");
		bindNameToId(legendGroupSelection);

		rectSelection = legendGroupSelection //
				.append("rect").onClick(this);

		updatePlotWithD3(d3);

		return legendGroupSelection;
	}

	@Override
	public void updatePlotWithD3(D3 d3) {
		plotPageModels(d3);
	}

	private void plotPageModels(D3 d3) {
		for (GraphicsPropertiesPageFactory pageModel : propertyPageFactories) {
			legendGroupSelection = pageModel.plotWithD3(d3, legendGroupSelection, rectSelection, this);
		}
	}

	//#end region

}

File: src/result/legend/legend.js
import PagedGraphicsAtom from './../graphics/pagedGraphicsAtom.js';
import Layout from './layout.js';
import Text from './text.js';
import Background from '../graphics/background.js';
import Border from '../graphics/border.js';

export default class Legend extends PagedGraphicsAtom {

	constructor(name){
		super(name);
		this.image = 'legend.png';		

		this.__legendGroupSelection = undefined;	
		this.__rectSelection = undefined;	

			
	}

	createPageFactories() {
		
		var factories = [];

		this.layout = Layout.create(this);
		factories.push(this.layout);

		this.text = Text.create(this);
		factories.push(this.text);

		this.background = Background.create(this);
		factories.push(this.background);

		this.border = Border.create(this);
		factories.push(this.border);
		
		return factories;
	}

	plot(dTreez, graphSelection, graphRectSelection, treeView) {
		
		this.treeView = treeView;

		graphSelection //
				.select('#' + this.name) //
				.remove();

		this.__legendGroupSelection = graphSelection //
				.append('g') //
				.className('legend');

        var dragSelection = this.__legendGroupSelection;
		var dragDeltaX = 0;
		var dragDeltaY = 0;

		let dragMethod = dTreez.drag()
							.on('start', () =>{
								dragDeltaX = this.layout.xTransform - dTreez.event.x;
								dragDeltaY = this.layout.yTransform - dTreez.event.y;

							})
							.on('drag', () => {
								let x = dTreez.event.x + dragDeltaX;
								let y = dTreez.event.y + dragDeltaY;
								this.layout.handleDragEvent(x, y);	
							});

		dragMethod(this.__legendGroupSelection);		
		
		this.bindString(()=>this.name, this.__legendGroupSelection, 'id');

		this.__rectSelection = this.__legendGroupSelection //
				.append('rect') //
				.onClick(()=>this.handleMouseClick());

		this.updatePlot(dTreez);

		return this.__legendGroupSelection;
	}
	
	updatePlot(dTreez) {
		this.__plotWithPages(dTreez);
	}

	__plotWithPages(dTreez) {
		for (var factory of this.__pageFactories) {
			factory.plot(dTreez, this.__legendGroupSelection, this.__rectSelection, this);
		}
	}	

}
